[visionOS] Don't unnecessarily recreate PlayableViewController if one still exists

jyavenard opened the pull request:
#### a6c9e44beeb48680d66936652d86a207f2476e6b
<pre>
[visionOS] Don&apos;t unnecessarily recreate PlayableViewController if one still exists
<a href="https://rdar.apple.com/139181732">rdar://139181732</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=282525">https://bugs.webkit.org/show_bug.cgi?id=282525</a>

Reviewed by Andy Estes.

Whenever we exited element fullscreen, VideoPresentationInterface calls invalidatePlayerViewController which clears its member m_playerViewController.
Immediately after, [WKFullscreenViewController configureEnvironmentPickerOfFullscreenVideoButtonView] is called which will have the VideoPresentationInterface
re-create the PlayableViewController. As a one step optimisation, have the LMK&apos;s `makeViewController` store the PlayableViewController created in a weakPtr
and subsequent call return the same value if it still exists.

No change in existing behaviour.

* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
(SwiftOnlyData.viewController):
(WKSLinearMediaPlayer.makeViewController):
* WebKitLibraries/SDKs/xros2.0-additions.sdk/System/Library/PrivateFrameworks/LinearMediaKit.framework/LinearMediaKit.tbd: Adding missing symbols
* WebKitLibraries/SDKs/xrsimulator2.0-additions.sdk/System/Library/PrivateFrameworks/LinearMediaKit.framework/LinearMediaKit.tbd: Adding missing symbols

Canonical link: <a href="https://commits.webkit.org/293400@main">https://commits.webkit.org/293400@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3bd34528c17ec711eb599700c44fb699bfd539e4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98837 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18474 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8714 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103963 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49426 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18769 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26924 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75246 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32380 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101841 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14259 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89264 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55606 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14050 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48806 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/83991 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7307 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106332 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25934 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18906 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84209 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26309 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85462 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83707 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21202 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28359 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/6030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19648 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25887 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/31073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25707 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29027 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27281 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
```diff
@@ -172,11 +173,14 @@ enum LinearMediaPlayerErrors: Error {
     func makeViewController() -> PlayableViewController {
         Logger.linearMediaPlayer.log("\(#function)")
 
+        if (swiftOnlyData.viewController != nil) { return swiftOnlyData.viewController! }
```
```suggestion
        if let viewController = swiftOnlyData.viewController {
            return viewController
        }
```

aestes reviewed (approved):
I think this is OK, but I do think there's a tradeoff here where you're optimizing one scenario at the expense of another. Exiting to element fullscreen is one case where I agree it's profitable to keep the view controller alive, but exiting to inline is another case where it seems mildly harmful to keep the view controller alive when the user likely won't re-enter fullscreen.

Ultimately r=me since I think element fullscreen is the common case and native (non-docked, non-immersive) video fullscreen is the edge case.

jyavenard committed:
[visionOS] Don't unnecessarily recreate PlayableViewController if one still exists
rdar://139181732
https://bugs.webkit.org/show_bug.cgi?id=282525

Reviewed by Andy Estes.

Whenever we exited element fullscreen, VideoPresentationInterface calls invalidatePlayerViewController which clears its member m_playerViewController.
Immediately after, [WKFullscreenViewController configureEnvironmentPickerOfFullscreenVideoButtonView] is called which will have the VideoPresentationInterface
re-create the PlayableViewController. As a one step optimisation, have the LMK's `makeViewController` store the PlayableViewController created in a weakPtr
and subsequent call return the same value if it still exists.

No change in existing behaviour.

* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
(SwiftOnlyData.viewController):
(WKSLinearMediaPlayer.makeViewController):
* WebKitLibraries/SDKs/xros2.0-additions.sdk/System/Library/PrivateFrameworks/LinearMediaKit.framework/LinearMediaKit.tbd: Adding missing symbols
* WebKitLibraries/SDKs/xrsimulator2.0-additions.sdk/System/Library/PrivateFrameworks/LinearMediaKit.framework/LinearMediaKit.tbd: Adding missing symbols

Canonical link: https://commits.webkit.org/293400@main
